src/System.Private.CoreLib/src/System/StartupHookProvider.cs

```
@@ -0,0 +1,90 @@
+                string typeName = assemblyPathAndTypeName[1];
+                if (String.IsNullOrEmpty(assemblyPath) || String.IsNullOrEmpty(typeName))
+                {
+                    throw new ArgumentException(SR.Argument_InvalidStartupHookSyntax);
+                }
+                startupHooksParsed.Add(Tuple.Create(assemblyPath, typeName));
+            }
+
+            // Ensure the startup dlls exist
+            foreach (var startupHook in startupHooksParsed)
+            {
+                var assemblyPath = startupHook.Item1;
+                if (!File.Exists(assemblyPath))
+                {
+                    throw new FileNotFoundException(SR.Format(SR.FileNotFound_ResolveAssembly, assemblyPath));
+                }
+            }
+
+            // Call each hook in turn
+            foreach (var startupHook in startupHooksParsed)
+            {
+                CallStartupHook(startupHook.Item1, startupHook.Item2);
+            }
+        }
+
+        // Load the specified assembly, and call the specified type's
+        // "public static void Initialize()" method.
+        private static void CallStartupHook(string assemblyPath, string typeName)
+        {
+            Debug.Assert(!String.IsNullOrEmpty(assemblyPath));
+            Debug.Assert(!String.IsNullOrEmpty(typeName));
+            Debug.Assert(File.Exists(assemblyPath));
+
+            string fullAssemblyPath = Path.GetFullPath(assemblyPath);
+            Assembly assembly = AssemblyLoadContext.Default.LoadFromAssemblyPath(fullAssemblyPath);
+            Debug.Assert(assembly != null);
+            Type type = assembly.GetType(typeName, throwOnError: true);
+            MethodInfo initializeMethod = type.GetMethod(initializeMethodName, Type.EmptyTypes);
+
+            if (initializeMethod == null)
+            {
+                throw new MissingMethodException(typeName, initializeMethodName);
+            }
+
+            Debug.Assert(initializeMethod.IsPublic);
```

I am intrigued by this assert. Since these hooks will most likely be used by people without a debug version of the SPCL, who is going to see this? If this is really a requirement, then I would throw an exception - which seems like a reasonable idea. I am all for asserts, but this isn't something anyone will do during development of this feature only during consumption.

---

This assert is dead code. `Type.GetMethod(String, Type[])` only looks at public methods.

---

I put the assert in there to document my expectation that GetMethod would only return a public method, so that it would fail fast should this not be the case during testing (in case I missed some edge case where that might not be true). We seem to use debug asserts similarly elsewhere in SPC, but I'm happy to remove it.